src/v/utils/tests/coro_bench.cc

```diff
@@ -0,0 +1,245 @@
+#include "base/seastarx.h"
+#include "ssx/future-util.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/testing/perf_tests.hh>
+#include <seastar/util/later.hh>
+
+#include <string>
+
+bool always_false_coro = false;
+
+namespace {
+
+static constexpr size_t ITERS = 10000;
+
+template<typename F>
+ss::future<size_t> co_await_in_loop(F f) {
+    perf_tests::start_measuring_time();
+    for (int i = 0; i < ITERS; i++) {
+        co_await f();
+    }
+    perf_tests::stop_measuring_time();
+    co_return ITERS;
+}
+
+template<typename F>
+ss::future<size_t> collect_futures(F f) {
+    std::vector<ss::future<>> futs;
+    futs.reserve(ITERS);
+    perf_tests::start_measuring_time();
+    for (int i = 0; i < ITERS; i++) {
+        futs.emplace_back(f());
+    }
+    perf_tests::stop_measuring_time();
+
+    for (auto& fut : futs) {
+        co_await std::move(fut);
+    }
+    co_return ITERS;
+}
+
+[[gnu::noinline]] ss::future<> ss_now() { return ss::now(); }
+
+[[gnu::noinline]] ss::future<> empty_cont() { return ss::maybe_yield(); }
+
+[[gnu::noinline]] static auto ready_future() {
+    return ss::make_ready_future<int>(1);
+}
+
+ss::future<int> empty_coro() {
+    int x = 0;
+    if (always_false_coro) {
+        x = co_await seastar::coroutine::without_preemption_check(
+          ready_future());
+    }
+    co_return x;
+}
+
+/**
+ * @brief Returns always ready future, but across an optimization barrier.
```

Comment:
Doesn't this need the no_inline attribute?

Reply:
The optimization barrier is the read of the `always_false_coro` and associated branch, not this function itself (I want this function to be inlined).